Add move_and_push and physics material presets

me-steven opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://contributing.godotengine.org/en/latest/engine/guidelines/code_style.html
-->

## Adds move_and_push() functions to CharacterBody3D and CharacterBody2D.
---------

### Reason: There are situations where you want players to push objects as they collide. 
------


- move_and_push() internally call move_and_slide() but applies an impulse to any rigid bodies that collides with the player

- Scaling with multiplier variable and PhysicsMaterial bounce value

- Adds collision multiplier to scale the impulse caused by the player

- Adds material presets to PhysicsMaterial

---

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/13211*

Ivorforce committed:
Complex body (#1)

* Expanded Physics Material

* Update physics_material.cpp

* Added move_and_push() for CharacterBody3D

- move_and_push() calls move_and_slide() under the hood and allows CharacterBody3D to push rigid bodies by applying impulse based on velocity between both and rigid bodies PhysicsMaterial

- Added Presets for PhysicsMaterial

* Added 2d move_and_push

-Added move and push to CharacterBody2D

-fixed typo in CharacterBody3D that would throw a debug message in editor. Typo didn't seem to affect how it worked.

added an error throw in case you manage to pass a physics material through the type check

* null check fix

-fixed a null check issue

* Added documentation

-added documentation
-minor tweaks to physics material preset values

Expanded Physics Material

Update physics_material.cpp

Physics Material Presets added - move and push added

added collision multiplier

Added 2d move_and_push

-Added move and push to CharacterBody2D

-fixed typo in CharacterBody3D that would throw a debug message in editor. Typo didn't seem to affect how it worked.

Added documentation

-added documentation
-minor tweaks to physics material preset values

STVND commented:
[2d example](https://youtu.be/wTOi8t6JIQg)

[3d example](https://youtu.be/iLEmpQn1iks)

AThousandShips commented on scene/resources/physics_material.cpp:
```diff
@@ -44,29 +47,112 @@ void PhysicsMaterial::_bind_methods() {
+			bounce = 0.2;
+			break;
+		default:
+			set_preset(PRESET_CUSTOM);
```
This will cause infinite recursion, did you test this code?

STVND replied:
I can remove the recursion but in testing the code I was unable to trigger it in anyway

me-steven committed:
Removed accidental recursion in Physics Material

Expanded Physics Material

Update physics_material.cpp

Added move_and_push() for CharacterBody3D

- move_and_push() calls move_and_slide() under the hood and allows CharacterBody3D to push rigid bodies by applying impulse based on velocity between both and rigid bodies PhysicsMaterial

- Added Presets for PhysicsMaterial

Added 2d move_and_push

-Added move and push to CharacterBody2D

-fixed typo in CharacterBody3D that would throw a debug message in editor. Typo didn't seem to affect how it worked.

formatting

Cleanup

null

Added documentation

-added documentation
-minor tweaks to physics material preset values

me-steven committed:
Merge branch 'master' of https://github.com/godotengine/godot

me-steven committed:
Create master.diff

me-steven committed:
Merge branch 'master' of https://github.com/STVND/godot

me-steven committed:
Adding upstream commits
